File: apps/expo-go/ios/Exponent/Versioned/Core/UniversalModules/EXNotifications/EXScopedNotificationSerializer.m
// Copyright 2018-present 650 Industries. All rights reserved.

#import "EXScopedNotificationSerializer.h"
#import "EXScopedNotificationsUtils.h"

NS_ASSUME_NONNULL_BEGIN

@implementation EXScopedNotificationSerializer

+ (NSDictionary<NSString *, NSObject *> *)serializedNotificationResponse:(UNNotificationResponse *)response
{
  NSDictionary *serializedResponse = [super serializedNotificationResponse:response];
  NSMutableDictionary *serializedResponseMutable = [serializedResponse mutableCopy];
  serializedResponseMutable[@"notification"] = [self serializedNotification:response.notification];
  
  return [serializedResponseMutable copy];
}

+ (NSDictionary<NSString *, NSObject *> *)serializedNotification:(UNNotification *)notification
{
  NSDictionary *serializedNotification = [super serializedNotification:notification];
  NSMutableDictionary *serializedNotificationMutable = [serializedNotification mutableCopy];
  serializedNotificationMutable[@"request"] = [self serializedNotificationRequest:notification.request];
  
  return [serializedNotificationMutable copy];
}

+ (NSDictionary<NSString *, NSObject *> *)serializedNotificationContent:(UNNotificationRequest *)request
{
  NSDictionary *serializedContent = [super serializedNotificationContent:request];
  NSMutableDictionary *serializedContentMutable = [serializedContent mutableCopy];
  serializedContentMutable[@"categoryIdentifier"] = request.content.categoryIdentifier ? [EXScopedNotificationsUtils getScopeAndIdentifierFromScopedIdentifier:request.content.categoryIdentifier].identifier : [NSNull null];
  
  return [serializedContentMutable copy];
}

+ (NSDictionary<NSString *, NSObject *> *)serializedNotificationRequest:(UNNotificationRequest *)request
{
  NSDictionary* serializedRequest = [super serializedNotificationRequest:request];
  NSMutableDictionary *serializedRequestMutable = [serializedRequest mutableCopy];
  serializedRequestMutable[@"identifier"] = [EXScopedNotificationsUtils getScopeAndIdentifierFromScopedIdentifier:request.identifier].identifier;

  return [serializedRequestMutable copy];
}

@end

NS_ASSUME_NONNULL_END

File: apps/expo-go/ios/Exponent/Versioned/Core/Modules/Notifications/EXScopedNotificationSerializer.swift
// Copyright 2018-present 650 Industries. All rights reserved.

import UserNotifications
import ExpoNotifications

public class EXScopedNotificationSerializer {

    public static func serializedNotification(_ notification: UNNotification) -> NotificationRecord {
        let scopedSerializedNotification = NotificationRecord(from: notification)
        scopedSerializedNotification.request = serializedNotificationRequest(notification.request)

        return scopedSerializedNotification
    }

    public static func serializedNotificationContent(_ request: UNNotificationRequest) -> NotificationContentRecord {
        let scopedSerializedContent = NotificationContentRecord(from: request)

        let categoryIdentifier = request.content.categoryIdentifier
        if (!categoryIdentifier.isEmpty) {
            let unscopedCategoryId = EXScopedNotificationsUtils.getScopeAndIdentifierFromScopedIdentifier(categoryIdentifier).identifier
            scopedSerializedContent.categoryIdentifier = unscopedCategoryId
        } else {
            scopedSerializedContent.categoryIdentifier = nil
        }

        return scopedSerializedContent
    }

    public static func serializedNotificationRequest(_ request: UNNotificationRequest) -> NotificationRequestRecord {
        let scopedSerializedRequest = NotificationRequestRecord(from: request)
        scopedSerializedRequest.identifier = EXScopedNotificationsUtils.getScopeAndIdentifierFromScopedIdentifier(request.identifier).identifier
        scopedSerializedRequest.content = serializedNotificationContent(request)

        return scopedSerializedRequest
    }
}
